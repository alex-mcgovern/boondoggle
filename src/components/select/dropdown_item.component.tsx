import clsx from "clsx";
import React, { forwardRef } from "react";

import { Box } from "../box";
import * as styles from "./select.styles.css";

import type { SharedUiScale } from "../..";
import type { DropdownItemShape } from "./select.types";
import type { Ref } from "react";

export interface DropdownItemProps {
  isHighlighted: boolean;
  isMultipleSelectionEnabled?: boolean;
  size?: SharedUiScale;
  isMulti?: boolean;
  item: DropdownItemShape;
  isDropdownItemSelected?: boolean;
}

export const DropdownItem = forwardRef(
  (
    {
      item,
      size,
      isHighlighted,
      isMulti,
      isDropdownItemSelected,
      ...rest
    }: DropdownItemProps,
    ref
  ) => {
    return (
      <Box
        as="button"
        id={item.label}
        className={clsx(
          styles.getDropdownItemStyles({
            size,
          }),
          {
            [styles.isHighlighted]: isHighlighted,
            [styles.isSelected]: isDropdownItemSelected,
          }
        )}
        {...rest}
        ref={ref as Ref<HTMLButtonElement>}
      >
        <Box flexShrink="0">{item.label}</Box>
        {isMulti && (
          <Box
            as="input"
            type="checkbox"
            readOnly
            tabIndex={-1}
            checked={isDropdownItemSelected}
            marginLeft="auto"
          />
        )}
      </Box>
    );
  }
);
