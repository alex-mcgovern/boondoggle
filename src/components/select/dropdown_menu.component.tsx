import { extractAtomsFromProps } from "@dessert-box/core";
import clsx from "clsx";
import React, { forwardRef } from "react";

import { getSprinkles } from "../../styles/utils/get_sprinkles.css";
import { Box } from "../box_component";
import { DropdownItem } from "./dropdown_item.component";
import * as styles from "./select.styles.css";

import type { SharedUiScale } from "../../styles/common/globalVariantsUiScale.css";
import type { SprinklesArgs } from "../../styles/utils/get_sprinkles.css";
import type { DropdownItemShape } from "./select.types";
import type {
  UseComboboxPropGetters,
  UseMultipleSelectionActions,
  UseMultipleSelectionGetSelectedItemPropsOptions,
} from "downshift";
import type { LegacyRef } from "react";

/**
 * ------------------------------------------------------------------------------
 * Util function to get props for individual dropdown items
 * May conditionally call `getItemProps` or `getSelectedItemProps` depending on
 * whether `isMulti` is true, and whether the item is selected or not
 * ------------------------------------------------------------------------------
 */

interface GetDropdownItemPropsArgs
  extends Pick<
    DropdownMenuProps,
    "getItemProps" | "getSelectedItemProps" | "removeSelectedItem"
  > {
  item: DropdownItemShape;
  isItemSelected?: boolean;
  [key: string]: unknown;
}

export const getDropdownItemProps = ({
  isItemSelected,
  getItemProps,
  getSelectedItemProps,
  removeSelectedItem,
  item,
  isMulti,
  ...rest
}: GetDropdownItemPropsArgs) => {
  if (isMulti && getSelectedItemProps) {
    return isItemSelected
      ? {
          ...getItemProps({
            item,
            ...rest,
          }),
          ...getSelectedItemProps({
            selectedItem: item,
            onClick: (e) => {
              e.stopPropagation();
              if (removeSelectedItem) {
                removeSelectedItem(item);
              }
            },
          }),
        }
      : getItemProps({ item, ...rest });
  }

  return getItemProps({ item });
};

/**
 * ------------------------------------------------------------------------------
 * Renders a dropdown menu for use with `SelectSingle` or `SelectMultiple`
 * ------------------------------------------------------------------------------
 */

export interface DropdownMenuProps extends SprinklesArgs {
  /** Required props  */
  getIsItemSelected: (item: DropdownItemShape) => boolean;
  getItemProps: UseComboboxPropGetters<DropdownItemShape>["getItemProps"];
  getMenuProps: UseComboboxPropGetters<DropdownItemShape>["getMenuProps"];
  isOpen: boolean;
  items: Array<DropdownItemShape>;

  /** Optional props  */
  getSelectedItemProps?: (
    options: UseMultipleSelectionGetSelectedItemPropsOptions<DropdownItemShape>
  ) => { [key: string]: unknown };
  isMulti?: boolean;
  highlightedIndex?: number | undefined;
  removeSelectedItem?: UseMultipleSelectionActions<DropdownItemShape>["removeSelectedItem"];
  size?: SharedUiScale;
}

export const DropdownMenu = forwardRef(
  (
    {
      getIsItemSelected,
      getItemProps,
      getMenuProps,
      getSelectedItemProps,
      highlightedIndex,
      removeSelectedItem,
      isMulti,
      isOpen,
      items,
      size = "md",
      ...rest
    }: DropdownMenuProps,
    ref
  ) => {
    const dropdownWrapperStyles = clsx(styles.dropdownListWrapper, {
      [styles.dropdownWrapperClosed]: !isOpen,
    });

    const { atomProps, otherProps } = extractAtomsFromProps(rest, getSprinkles);

    /**
     * Note: `DropdownMenu` *must* not be in a conditional render, or
     * downshift's `getMenuProps` will be unable to apply a ref and throw an error
     */
    return (
      <Box
        className={dropdownWrapperStyles}
        {...atomProps}
        {...getMenuProps?.({
          ...otherProps,
          ref: ref as LegacyRef<HTMLElement>,
        })}
      >
        {Array.isArray(items) &&
          items.map((item, index) => {
            if (!item.label) {
              return null;
            }

            const isItemSelected = getIsItemSelected?.(item);

            return (
              <DropdownItem
                item={item}
                isMulti={isMulti}
                isHighlighted={highlightedIndex === index}
                key={item.label}
                isDropdownItemSelected={isItemSelected}
                size={size}
                {...getDropdownItemProps({
                  isItemSelected,
                  index,
                  getItemProps,
                  getSelectedItemProps,
                  item,
                  isMulti,
                  removeSelectedItem,
                })}
              />
            );
          })}
      </Box>
    );
  }
);
