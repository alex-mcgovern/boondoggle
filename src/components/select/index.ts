export { SelectSingle } from "./select_single.comp";
export type { SelectSingleProps } from "./select_single.comp";

export { SelectButton } from "./select_button.comp";
export type { SelectButtonProps } from "./select_button.comp";

export { SelectMulti } from "./select_multi.comp";
export type { SelectMultiProps } from "./select_multi.comp";

export type { DropdownItemShape } from "./select.types";
