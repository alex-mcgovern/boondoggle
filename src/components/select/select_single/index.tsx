import { autoUpdate, flip, offset, useFloating } from "@floating-ui/react";
import { faAngleDown } from "@fortawesome/sharp-regular-svg-icons";
import { useCombobox } from "downshift";
import { forwardRef, useCallback, useState } from "react";

import { getOptionalLabelProps } from "../../../common-types";
import { useForwardRef } from "../../../hooks/use_forward_ref";
import { Box } from "../../box";
import { Icon } from "../../icon";
import { Input } from "../../input";
import { filterSelectItems } from "../lib/filter_select_items";
import { getSlotRight } from "../lib/get_slot_right";
import { SelectItemList } from "../select_item_list";
import { getIsSelected } from "../select_utils";
import { selectInputCursorStyles } from "../shared/select_input.styles.css";

import type {
    WithId,
    WithName,
    WithOptionalIsClearable,
    WithOptionalLabel,
    WithOptionalPlaceholder,
    WithSize,
    WithSlots,
    WithStateDisabled,
    WithStateInvalid,
    WithWrapperProps,
} from "../../../common-types";
import type { InputProps } from "../../input";
import type { SelectItemShape } from "../types";
import type { UseComboboxStateChange } from "downshift";

export type SelectSingleProps = Omit<WithOptionalIsClearable, "readOnly"> &
    WithName &
    WithOptionalPlaceholder &
    WithSize &
    WithSlots &
    WithStateDisabled &
    WithStateInvalid &
    WithWrapperProps &
    WithId &
    WithOptionalLabel & {
        /**
         * Item to be preselected when the component mounts.
         */
        initialSelectedItem?: SelectItemShape;

        /**
         * Props to customise the input element.
         */
        inputProps?: Partial<
            Omit<
                InputProps,
                | "id"
                | "isClearable"
                | "isCopyable"
                | "isVisibilityToggleable"
                | "isVisible"
                | "label"
                | "labelProps"
                | "size"
                | "labelTooltip"
                | "width"
            >
        >;

        /**
         * Whether the Select should be filterable by typing.
         */
        isFilterable?: boolean;

        /**
         * Prop to toggle the open state of the dropdown.
         */
        isOpen?: boolean;

        /**
         * Function to convert an item to a string.
         */
        itemToString?: (item: SelectItemShape | null) => string;

        /**
         * The items to render in the dropdown.
         */
        items: Array<SelectItemShape>;

        /**
         * Function called with the new selected item when the selection changes.
         */
        onChange?: (selection: SelectItemShape | null | undefined) => void;

        /**
         * Function called with the new open state when the dropdown is opened or closed.
         */
        onIsOpenChange?: (changes: UseComboboxStateChange<SelectItemShape>) => void;
    };

export const SelectSingle = forwardRef<HTMLInputElement, SelectSingleProps>(
    (
        {
            disabled,
            errorMessage,
            id,
            initialSelectedItem,
            inputProps,
            invalid,
            isClearable,
            isFilterable,
            isOpen: controlledIsOpen,
            itemToString = (item: SelectItemShape | null) => {
                return item?.label || "";
            },
            items: initialItems,
            label,
            labelTooltip,
            name,
            onChange,
            onIsOpenChange,
            placeholder,
            size,
            slotLeft,
            slotRight = [<Icon icon={faAngleDown} />],
            wrapperProps,
        },
        initialRef
    ) => {
        const ref = useForwardRef(initialRef);

        const initialItem =
            initialSelectedItem ||
            initialItems.find((item) => {
                return item.isSelected;
            });

        const [inputValue, setInputValue] = useState(initialItem?.label || "");

        const items = isFilterable
            ? filterSelectItems({ inputValue, items: initialItems })
            : initialItems;

        const {
            getInputProps,
            getItemProps,
            getLabelProps,
            getMenuProps,
            highlightedIndex,
            isOpen,
            reset,
            selectedItem,
        } = useCombobox({
            initialSelectedItem: initialItem,
            inputValue,
            isItemDisabled: (item) => {
                return item.disabled;
            },
            isOpen: controlledIsOpen,
            items,
            itemToString,
            onInputValueChange: (changes) => {
                setInputValue(changes.inputValue || "");
            },
            onIsOpenChange,
            onSelectedItemChange: (changes) => {
                return onChange?.(changes.selectedItem);
            },
            // Ensure that onClick is called when the user presses Enter on an item.
            onStateChange(changes) {
                if (changes.type === useCombobox.stateChangeTypes.InputKeyDownEnter) {
                    changes.selectedItem?.onClick?.();
                }
            },
        });

        const getIsItemSelected = useCallback(
            (item: SelectItemShape) => {
                return getIsSelected({
                    item,
                    selectedItem,
                });
            },
            [selectedItem]
        );

        const { floatingStyles, refs } = useFloating({
            elements: {
                reference: ref.current,
            },
            middleware: [
                offset(4),
                flip({
                    crossAxis: true,
                    fallbackAxisSideDirection: "start",
                }),
            ],
            open: isOpen,
            placement: "bottom-start",
            whileElementsMounted: autoUpdate,
        });

        return (
            <Box
                position="relative"
                {...wrapperProps}
            >
                <Input
                    errorMessage={errorMessage}
                    size={size}
                    slotLeft={selectedItem?.slotLeft || slotLeft}
                    slotRight={getSlotRight({
                        isClearable:
                            (!!isFilterable && !!inputValue) || (!!isClearable && !!selectedItem),
                        reset,
                        slotRight,
                    })}
                    {...getInputProps({
                        className: selectInputCursorStyles,
                        disabled,
                        id,
                        invalid,
                        isClearable: undefined,
                        name,
                        placeholder,
                        readOnly: !isFilterable,
                        ref,
                        value: inputValue,
                        ...getOptionalLabelProps({
                            id,
                            label,
                            labelProps: getLabelProps({
                                htmlFor: id,
                            }),
                            labelTooltip,
                        }),
                        ...inputProps,
                    })}
                />

                <SelectItemList
                    getIsItemSelected={getIsItemSelected}
                    getItemProps={getItemProps}
                    getMenuProps={getMenuProps}
                    getSelectedItemProps={undefined}
                    highlightedIndex={highlightedIndex}
                    isMulti={false}
                    isOpen={isOpen}
                    items={items}
                    ref={isOpen ? refs.setFloating : undefined}
                    size={size}
                    style={floatingStyles}
                />
            </Box>
        );
    }
);
