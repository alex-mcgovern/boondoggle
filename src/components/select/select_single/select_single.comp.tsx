import { useCombobox } from "downshift";
import { forwardRef, useCallback, useMemo, useState } from "react";

import { Box } from "../../box";
import { Input } from "../../input";
import {
  downshiftStateReducer,
  getDefaultHighlightedIndex,
  getFilteredDropdownItems,
  getIsSelected,
} from "../select_utils";
import { DEFAULT_SLOT_RIGHT } from "../shared/DEFAULT_SLOT_RIGHT";
import { DropdownMenu } from "../shared/dropdown_menu/dropdown_menu.comp";
import { selectInputCursorStyles } from "../shared/select_input.styles.css";

import type { LabelledElementCustomisation } from "../../../types";
import type { DropdownItemShape, SelectCommonProps } from "../select.types";
import type { UseComboboxStateChange } from "downshift";
import type { Ref } from "react";

/**
 * Util function for transforming a dropdown item into a string (for use as a `value` for {@link Input}).
 * Can be overridden by passing a custom `itemToString` prop to the `SelectSingle` component.
 */
const defaultItemToString = (item: DropdownItemShape | null) => {
  return item?.label || "";
};

export type SelectSingleProps = SelectCommonProps &
  LabelledElementCustomisation & {
    initialSelectedItem?: DropdownItemShape | null;
    itemToString?: (item: DropdownItemShape | null) => string;
    /** Optional tooltip for label */
    labelTooltip?: string;
    onChange?: (changes: UseComboboxStateChange<DropdownItemShape>) => void;
    placeholder: string;
  };

/** Accessible select component, supports multi & single modes. */
export const SelectSingle = forwardRef(
  (
    {
      disabled,
      errorMessage,
      id,
      initialHighlightedItem,
      initialSelectedItem,
      invalid,
      isFilterable,
      items,
      itemToString = defaultItemToString,
      label,
      name,
      onChange,
      labelTooltip,
      placeholder,
      size,
      slotLeft,
      slotRight = DEFAULT_SLOT_RIGHT,
      wrapperProps,
      ...rest
    }: SelectSingleProps,
    ref: Ref<HTMLInputElement>
  ) => {
    const [selectedItem, setSelectedItem] = useState(
      initialSelectedItem || null
    );

    const [inputValue, setInputValue] = useState(
      initialSelectedItem?.label || ""
    );

    const [localSlotLeft, setLocalSlotLeft] = useState(slotLeft);

    /** Filter dropdown items based on input if `isFilterable` is true */
    const filteredItems = useMemo(() => {
      if (!items || !isFilterable) {
        return items;
      }

      return getFilteredDropdownItems({ inputValue, items });
    }, [items, isFilterable, inputValue]);

    /** Initialise downshift `useCombobox` hook */
    const {
      getInputProps,
      getItemProps,
      getMenuProps,
      getLabelProps,
      highlightedIndex,

      isOpen,
      toggleMenu,
    } = useCombobox({
      defaultHighlightedIndex: getDefaultHighlightedIndex({
        initialHighlightedItem,
        items,
        selectedItem,
      }),
      initialSelectedItem,
      items: filteredItems,
      itemToString,
      onSelectedItemChange: onChange,
      onStateChange({
        inputValue: newInputValue,
        type,
        selectedItem: newSelectedItem,
      }) {
        switch (type) {
          case useCombobox.stateChangeTypes.InputKeyDownEnter:
          case useCombobox.stateChangeTypes.ItemClick:
          case useCombobox.stateChangeTypes.InputBlur:
            if (newSelectedItem) {
              setSelectedItem(newSelectedItem);
              setInputValue(newSelectedItem.label);
              if (newSelectedItem.slotLeft) {
                setLocalSlotLeft(newSelectedItem.slotLeft);
              }
            }
            break;

          case useCombobox.stateChangeTypes.InputChange:
            if (typeof newInputValue !== "undefined") {
              setInputValue(newInputValue);
            }

            break;
          default:
            break;
        }
      },
      selectedItem,
      stateReducer: (state, actionAndChanges) => {
        return downshiftStateReducer(state, actionAndChanges, {});
      },
    });

    const getIsItemSelected = useCallback(
      (item: DropdownItemShape) => {
        return getIsSelected({
          item,
          selectedItem,
        });
      },
      [selectedItem]
    );

    /** -----------------------------------------------------------------------------
     * Layout for select component
     * ------------------------------------------------------------------------------- */

    return (
      <Box {...wrapperProps}>
        <DropdownMenu
          getIsItemSelected={getIsItemSelected}
          getItemProps={getItemProps}
          getMenuProps={getMenuProps}
          highlightedIndex={highlightedIndex}
          isOpen={isOpen}
          items={filteredItems}
          size={size}
          width="100%"
          triggerNode={
            // eslint-disable-next-line react-perf/jsx-no-jsx-as-prop
            <Input
              className={selectInputCursorStyles}
              errorMessage={errorMessage}
              invalid={invalid}
              label={label}
              labelTooltip={labelTooltip}
              readOnly={!isFilterable}
              size={size}
              slotLeft={localSlotLeft}
              slotRight={slotRight}
              labelProps={getLabelProps({
                htmlFor: id,
              })}
              {...getInputProps?.({
                disabled,
                id,
                name,
                onClick: toggleMenu,
                placeholder,
                ref,
                value: inputValue,
                ...rest,
              })}
            />
          }
        />
      </Box>
    );
  }
);
