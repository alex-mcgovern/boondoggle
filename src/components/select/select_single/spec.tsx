import { SelectSingle } from ".";
import { LOREM } from "../../../../mocks/LOREM.mock";
import { mockSelectItems } from "../__mocks__/select.mock";

describe("<SelectSingle />", () => {
    it("mounts", () => {
        cy.mount(
            <SelectSingle
                {...{
                    id: LOREM.id(),
                    items: mockSelectItems({}),
                    label: LOREM.label(),
                    name: LOREM.name(),
                    onChange: (selection) => {
                        alert(`onChange\n\n${JSON.stringify(selection, null, 2)}`);
                    },
                    placeholder: LOREM.placeholder(),
                }}
            />
        );

        cy.get(`input[name='${LOREM.name()}']`).click();
        cy.get(`li[value=${mockSelectItems({})[0].value}]`).click();
    });
});
