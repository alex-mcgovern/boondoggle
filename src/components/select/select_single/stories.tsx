/* eslint-disable import/no-extraneous-dependencies */
import { faGlobeAfrica } from "@fortawesome/sharp-regular-svg-icons";

import { SelectSingle as StoryComp } from ".";
import { LOREM } from "../../../../mocks/LOREM.mock";
import { Icon } from "../../icon";
import { mockSelectItems } from "../__mocks__/select.mock";

import type { Meta, StoryObj } from "@storybook/react";

const meta = {
    args: {
        id: LOREM.id(),
        items: mockSelectItems({}),
        label: LOREM.label(),
        name: LOREM.name(),
        onChange: (selection) => {
            alert(`onChange\n\n${JSON.stringify(selection, null, 2)}`);
        },
        placeholder: LOREM.placeholder(),
    },
    component: StoryComp,
    title: "Components/SelectSingle",
} satisfies Meta<typeof StoryComp>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {};

export const InitialSelectedItem: Story = {
    args: {
        initialSelectedItem: mockSelectItems({})[0],
    },
};

export const Invalid: Story = {
    args: {
        errorMessage: LOREM.errorMessage(),
        invalid: true,
    },
};

export const IsClearable: Story = {
    args: {
        isClearable: true,
    },
};

export const IsFilterable: Story = {
    args: {
        isFilterable: true,
    },
};

export const Customisation: Story = {
    args: {
        wrapperProps: {
            background: "tint_default",
            border: "border_default",
            borderRadius: "lg",
            padding: "space_5",
            // width: "100%",
        },
    },
};

export const Disabled: Story = {
    args: {
        disabled: true,
    },
};

export const SizeSm: Story = {
    args: {
        size: "sm",
    },
};

export const SizeMd: Story = {
    args: {
        size: "md",
    },
};

export const SizeLg: Story = {
    args: {
        size: "lg",
    },
};

export const WithButtonsAsItems: Story = {
    args: {
        items: mockSelectItems({
            onClick: () => alert("clicked"),
        }),
    },
};

export const WithItemsWithSelectedTrue: Story = {
    args: {
        items: [
            {
                isSelected: true,
                label: "United Kingdom",
                value: "uk",
            },
            {
                isSelected: false,
                label: "France",
                value: "fr",
            },
            {
                isSelected: false,
                label: "Germany",
                value: "de",
            },
        ],
    },
};

export const WithItemsWithSlotLeft: Story = {
    args: {
        items: mockSelectItems({ withIcon: true }),
    },
};

export const WithSlotLeft: Story = {
    args: {
        slotLeft: [<Icon icon={faGlobeAfrica} />],
    },
};

export const WithSlots: Story = {
    args: {
        slotRight: [<Icon icon={faGlobeAfrica} />],
    },
};
