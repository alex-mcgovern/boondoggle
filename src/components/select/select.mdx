import { ArgTypes, Canvas, Meta, Story } from "@storybook/addon-docs";

import * as stories from "./select.stories.tsx";

<Meta of={stories} />

# Select

- Wrapper around Downshift to provide a select component with a few options for
  styling and behavior

## Select Single

<Canvas name="default" of={stories.SelectSingleDefault} />

## Select multiple

<Canvas name="default" of={stories.SelectMultipleDefault} />

## Filterable

[ToDo: Add filterable example]
{/* <Canvas name="filterable" of={stories.Filterable} /> */}
