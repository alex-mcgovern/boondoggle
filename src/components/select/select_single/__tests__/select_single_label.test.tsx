/** @jest-environment jsdom */
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { LOREM } from "../../../../../mocks/LOREM.mock";
import "../../../../../test/dialog.mock";
import { mockSelectItems } from "../../__mocks__/select.mock";
import { SelectSingle } from "../select_single.comp";
import { SelectSingleProps } from "../select_single.comp";

/** -----------------------------------------------------------------------------
 * Test setup
 * ------------------------------------------------------------------------------- */

const PROPS: SelectSingleProps = {
  id: LOREM.id(),
  items: mockSelectItems({}),
  label: LOREM.label(),
  name: LOREM.textXxs,
  placeholder: LOREM.select,
};

const renderComponent = (props: SelectSingleProps) => {
  return {
    user: userEvent.setup(),
    ...render(<SelectSingle {...props} />),
  };
};

/** -----------------------------------------------------------------------------
 * Test cases
 * ------------------------------------------------------------------------------- */

describe("<SelectSingle />", () => {
  it("should be labelled", async () => {
    const { container, getByRole } = await renderComponent(PROPS);

    const combobox = getByRole("combobox");
    const label = container.querySelector("label");
    const labelId = label?.getAttribute("id");

    expect(combobox.getAttribute("aria-labelledby")).toBe(labelId);
    expect(label).not.toBeNull();
    expect(label?.getAttribute("for")).toBe(PROPS.id);
    expect(label?.textContent).toBe(PROPS.label);
  });

  it("should match snapshot", async () => {
    const { getByRole } = await renderComponent(PROPS);

    expect(getByRole("combobox")).toMatchSnapshot();
  });
});
