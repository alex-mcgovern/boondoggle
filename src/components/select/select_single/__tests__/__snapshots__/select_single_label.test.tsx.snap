// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectSingle /> should match snapshot 1`] = `
<input
  aria-activedescendant=""
  aria-autocomplete="list"
  aria-controls="downshift-2-menu"
  aria-expanded="false"
  aria-labelledby="downshift-2-label"
  autocomplete="off"
  class="slot_wrapper_inset.styles_offsetSlotRight__1r5w8or2 input.styles_getInputStyles__3x32wp4 input.styles__3x32wp3 get_sprinkles_width_100%_mobile__eiij7bda a11y_a11yDisabled__r8he5t1 input.styles_getInputStyles_size_md__3x32wp6 input.styles_size_md__3x32wp1 element_size_elementFontSize_md__1f6zrv4 element_size__1f6zrv1 get_sprinkles_fontSize_body_md__eiij7b28 get_sprinkles_lineHeight_body_md__eiij7b3w element_size_elementPadding_md__1f6zrva element_size__1f6zrv7 get_sprinkles_paddingLeft_spacing1.5__eiij7b9c get_sprinkles_paddingRight_spacing1.5__eiij7b9n get_sprinkles_paddingTop_spacing1__eiij7b9x get_sprinkles_paddingBottom_spacing1__eiij7b90 select_input.styles_selectInputCursorStyles__1bkra5a0"
  id="lorem-ipsum"
  name="Lorem ipsum"
  placeholder="Select a thing"
  readonly=""
  role="combobox"
  value=""
/>
`;
